### ISSUE TITLE
RL Stack Refactoring: Call for Contributions

### ISSUE THREAD

**Issue by s1lent4gnt** [huggingface/lerobot#3076]
## RL Stack Refactoring — Call for Contributions

We have been working on a major RL stack refactoring in LeRobot. The goal is to build a **solid RL foundation** that makes reinforcement learning and fine-tuning VLAs with RL as easy as imitation learning already is — and to make **adding new RL algorithms** straightforward for the community.

The first RL algorithm in LeRobot — **HIL-SERL** (RLPD/SAC with actor-learner architecture, reward classifier, and human interventions) — is working ([#504](https://github.com/huggingface/lerobot/issues/504), [PR #644](https://github.com/huggingface/lerobot/pull/644)). Now we want to:
1. **Solidify the foundations** and tighten the interfaces
2. **Make adding new RL algorithms easy** — one algorithm, one file
3. **Support VLA fine-tuning with RL** (ConRFT, QC-FQL, RECAP, etc)
4. **Integrate pluggable reward models** for both IL and RL — see [#3143](https://github.com/huggingface/lerobot/issues/3143), [PR #3142](https://github.com/huggingface/lerobot/pull/3142)

We welcome community feedback, contributions, and ideas. You don't need to write extensive code on your own — any input on a sub-component or comment, however small, is appreciated.

**Coordination:** Discord [#reinforcement-learning](https://discord.com/channels/1216765309076115607/1304507548204007576)

---

## Current Architecture

This architecture is introduced in the [RL refactoring PR](https://github.com/huggingface/lerobot/pull/3075). The RL stack lives in `src/lerobot/rl/` and follows LeRobot's existing patterns:

```
src/lerobot/rl/
├── algorithms/
│   ├── base.py          # RLAlgorithm ABC + RLAlgorithmConfig (draccus.ChoiceRegistry)
│   └── sac.py           # SAC training implementation
├── trainer.py           # RLTrainer (orchestrates training steps)
├── buffer.py            # ReplayBuffer
├── data_sources/
│   └── data_mixer.py    # DataMixer + OnlineOfflineMixer
├── learner.py           # Learner process (GPU training)
├── actor.py             # Actor process (environment interaction)
├── learner_service.py   # gRPC service for distributed training
└── ...
```

**Key design decisions already in place:**
- **`RLAlgorithmConfig`** uses `draccus.ChoiceRegistry` — same pattern as policies, cameras, motors
- **`RLAlgorithm.update(batch_iterator)`** — algorithms consume from an iterator, controlling batch consumption (e.g. UTD ratio)
- **Actor-Learner architecture** — distributed via gRPC, actors collect experience, learner trains
- **`DataMixer`** — mixes heterogeneous data sources (OnlineOfflineMixer mixes offline + online demonstrations)

---

### Phase 1: Solidify Foundations (SAC stays working, interfaces tighten)
- [ ] Make `RLAlgorithmConfig` a first-class explicit config (decouple from `policy_cfg`)
- [ ] Clean up `RLAlgorithm` public API (`select_action`, `update`, `get_weights`/`load_weights`, `configure_data_iterator`)
- [ ] Improve documentation and add an RL training tutorial
- [ ] Add tests for the core RL components
- [ ] Benchmarks comparing RL methods

### Phase 2: Pluggable Reward Models
- [ ] Create `src/lerobot/rewards/` module — shared reward models for both IL and RL
- [ ] Migrate existing reward classifier from `policies/sac/reward_model/`
- [ ] Migrate SARM from `policies/sarm/` (see [SampleWeighter refactor](https://github.com/huggingface/lerobot/pull/2781) for related IL abstraction)
- [ ] Support zero-shot reward models (TOPReward, VITA)
- [ ] **Papers:** [RoboReward](https://arxiv.org/abs/2601.00675), [VITA](https://arxiv.org/abs/2506.10085)

### Phase 3: VLA Fine-Tuning Algorithms
- [ ] **RECAP** — Advantage-conditioned policies for heterogeneous data. Paper: [π*0.6](https://arxiv.org/abs/2511.14759). Community: [lerobot_recap](https://github.com/cijerezg/lerobot/tree/my-pi05-merge), [PR #2923](https://github.com/huggingface/lerobot/pull/2923)
- [ ] **QC_FQL** - Flow Q-learning (FQL) agent with action chunking (QC). Community: [PR #1818](https://github.com/huggingface/lerobot/pull/1818)
- [ ] **ConRFT** — Unified offline-online consistency policy. Community: [PR #1823](https://github.com/huggingface/lerobot/pull/1832)
- [ ] **DSRL** (low complexity) — Steer frozen diffusion policy via latent-noise SAC; builds on existing SAC infra. Paper: [arXiv:2506.15799](https://arxiv.org/abs/2506.15799) — LeRobot branch by @michel-aractingi [DSRL](https://github.com/huggingface/lerobot/tree/feat/add_dsrl)
- [ ] **TwinRL** — Paper: [arXiv:2602.09023](https://arxiv.org/abs/2602.09023)

---

## Open for Ideas

We want to hear from the community:
- What would make your life easier when training RL policies in LeRobot?
- What RL algorithms are you most excited about for robot learning?
- What other RL-based VLA fine-tuning methods should we support?
- What reward model approaches have worked best for your tasks?

Please **comment below** or **open a PR**. Contributions are highly encouraged!

Your help will make LeRobot a powerful and accessible framework for robot RL 🤖

**Comment by neo-alex** [comment]
Hey @s1lent4gnt, this RL refactoring sounds like a very exciting improvement - especially the support for VLA fine-tuning with RL. Really looking forward to it 👍 

> What would make your life easier when training RL policies in LeRobot?

--> I think it is already part of your roadmap, but what would be really great is to decouple as much as possible the RL algorithms (e.g. SAC or its fancier HIL-SERL version) from their policy/critic neural net definition: for instance, when training HIL-SERL, this would easily enable reusing a (pretrained) VLA of our choice instead of starting from scratch with a hardcoded model [like today](https://github.com/huggingface/lerobot/blob/85de893fa73fa939f1d57d209b2c04be11a0a71e/src/lerobot/policies/sac/modeling_sac.py#L449).

Thanks!

**Comment by sezan92** [comment]
Really interested to start working on it

**Comment by s1lent4gnt** [comment]
Hey @neo-alex, thanks for the feedback!

> --> I think it is already part of your roadmap, but what would be really great is to decouple as much as possible the RL algorithms (e.g. SAC or its fancier HIL-SERL version) from their policy/critic neural net definition: for instance, when training HIL-SERL, this would easily enable reusing a (pretrained) VLA of our choice instead of starting from scratch with a hardcoded model [like today](https://github.com/huggingface/lerobot/blob/85de893fa73fa939f1d57d209b2c04be11a0a71e/src/lerobot/policies/sac/modeling_sac.py#L449).

Yes this is in our roadmap.

**Comment by jackvial** [comment]
> What would make your life easier when training RL policies in LeRobot?

EDIT/TL;DR - I think I was getting a bit ahead of myself here and being over specific without having enough context yet but I still agree with my general sentiment of having a more incremental introduction to RL starting with something like simple inference in sim with a ready to run pretrained policy, then building up to more advanced methods like HIL on real hardware and zero shot sim2real. Maybe there would be multiple tracks RL sim2real vs real only methods like RECAP.

---

@s1lent4gnt Improved documentation and first class support for so101 in sim and on real hardware would help a lot, here's some of notes on the current RL docs and some suggestions for improvements that I think might make getting up and running with RL in LeRobot a bit a smoother.

## High-Level Feedback

The current RL documentation is centered around **HIL-SERL**, I think  HIL-SERL should instead be treated as a special case of RL rather than the primary entry point.

More concrete examples in the docs and following more of a tiered progression could help a lot. Users should be able to start with a minimal working example and gradually move toward more advanced setups like deploying on real hardware and HIL training.

---

# Proposed RL Documentation Structure

## 1. Run a Pretrained RL Policy in Simulation

Provide the **simplest possible entry point**.

Goal:
- Run a pretrained RL policy for an *SO101 arm in simulation*
- Should ideally be **a single command**
- Get quick feedback and confirm that the environment and dependencies are set up correctly for inference

Example workflow:
- Download pretrained model
- Launch simulator
- Run inference

This step should require **no training or configuration**.

Purpose:
- Fast early feedback and context setting
- Reduce onboarding friction
- Validate environment setup

---

## 2. Train an RL Policy in Simulation

Next step should introduce basic RL training in simulation.

Suggested configuration:

- **Robot:** SO101  
- **Algorithm:** Squint  
- **Task:** Lift Cube  
- **Simulator:** ManiSkill  

Reasoning:

- Squint is very fast to train **(~15 minutes on RTX 4070)**.
- The Squint method helps reduce the sim2real gap and the paper reports success with zero shot deployment on real hardware
- so101 + Manisill + lift cube because it's tried and tested with squint in https://github.com/aalmuzairee/squint 

Recommended walkthrough:

1. Train squint for so101 + lift cube environment
2. Evaluate trained policy in sim

---

## 3. Train an RL Model for Deployment on Real Hardware

This stage introduces sim2real deployment on real hardware.

Use the **same setup as Step 2** to maintain continuity:

- Robot: SO101
- Algorithm: Squint
- Task: Lift Cube
- Simulator: ManiSkill

Primary focus should be on the sim2real gap challenges.

Topics to cover:

### Camera Calibration

Provide tooling and walkthroughs for calibration.

Relevant tools:
- Manual camera calibration  https://github.com/aalmuzairee/squint/blob/master/deploy_utils/tune_camera.py
- Would be good to consider EasyHEC / SimplyEasyHEC  https://github.com/StoneT2000/simple-easyhec

### Background Masking
- Include instructions on creating and configuring the background mask to help close the sim2real gap

## My Experience with Squint and Sim2Real

The Squint paper reports zero-shot sim-to-real success, I have trained squint and deployed it on my so101, this is what I found:
- Model training was fast and the general process was quite smooth aside from camera calibration
- My camera alignment was imperfect and this is still quite challenging
- I did not get the robot to complete the task successfully but the motion was not bad and I feel like I got a lot further quicker and with better results than with PPO

---

## 4. HIL RL Training on SO101

Since Squint is SAC.based it should be possible to extend it for HIL (Human-in-the-Loop) RL training.

If so this would make for a natural progression from Step 3.

Topics to cover:

1. Setting up the reward classifier
3. HIL training

---

## 5. RL Training for VLA Models

I need to get more familiar with RL for VLA models so not going to say much on it atm but it seems like it would make sense as the next step in the progression

**Comment by lvjonok** [comment]
Hello there!

I am very happy that RL stack will be modernized throughout LeRobot. I would personally benefit a lot from this RL stack not being too fine-tuned for the SO101 robot. Though it is the one people have access the most, but I believe many labs nowadays use some kind of a wrapper around LeRobot to use with their own robots like Franka, UR and so on. Thus, having LeRobot be agile to embodiment (or at least try not to use something outside of the base Robot class) will simplify the integration for the whole community.

**Comment by sezan92** [comment]
@s1lent4gnt i want to start contributing in this issue, can you give me guidance where to start , who to talk?

**Comment by tourin1992** [comment]
Same here! Would love to contribute! Please let me know where to start :) @s1lent4gnt

**Comment by s1lent4gnt** [comment]
> Hello there!
> 
> I am very happy that RL stack will be modernized throughout LeRobot. I would personally benefit a lot from this RL stack not being too fine-tuned for the SO101 robot. Though it is the one people have access the most, but I believe many labs nowadays use some kind of a wrapper around LeRobot to use with their own robots like Franka, UR and so on. Thus, having LeRobot be agile to embodiment (or at least try not to use something outside of the base Robot class) will simplify the integration for the whole community.

@lvjonok thanks for raising this, that’s definitely an important point. Keeping the RL stack robot-agnostic is definitely one of the goals of the refactor.

**Comment by s1lent4gnt** [comment]
Hey @sezan92 and @tourin1992, thank you for your interest in contributing!

It depends on what you're interested in working on, it could be a new RL algorithm like DSRL, TwinRL, or others from the roadmap, or it could be on the infrastructure side (tests, documentation, reward models, etc.).

Our main goal is to make RL in LeRobot first-class and as simple as imitation learning already is. So the priority right now is to solidify the current implementation and tighten the interfaces to make sure adding, training, and evaluating new RL algorithms is straightforward for everyone.

For coordination, feel free to join the [**#reinforcement-learning**](https://discord.com/channels/1216765309076115607/1304507548204007576) or [**#dev-contributing**](https://discord.com/channels/1216765309076115607/1457823929719722087) channel on Discord as well!

**Comment by TaherTadpatri** [comment]
Not able to join discord.

**Comment by s1lent4gnt** [comment]
> Not able to join discord.

Hey @TaherTadpatri, you can join here : https://discord.com/invite/s3KuuzsPFb

**Comment by sezan92** [comment]
@s1lent4gnt i would like to work on DSRL implementation, please let me know if there is any github issue for it, if not i will make an issue

**Comment by s1lent4gnt** [comment]
> [@s1lent4gnt](https://github.com/s1lent4gnt) i would like to work on DSRL implementation, please let me know if there is any github issue for it, if not i will make an issue

Hey @sezan92 — great! You can start from Michel’s previous work here:
https://github.com/huggingface/lerobot/tree/feat/add_dsrl
https://github.com/huggingface/lerobot/tree/tmp/wip_dsrl

Also, please check the original paper and code:
https://diffusion-steering.github.io/

**Comment by sezan92** [comment]
@s1lent4gnt  thanks , sorry for late reply, i will reply

**Comment by sezan92** [comment]
@s1lent4gnt  i have an update, i started working little by littke. i think first task for me is to refactor learner.py into learner.py and a separate sac.py. i am writing an issue about it. later i will add dsrl.py to this so the learner.py can access it.

**Comment by sezan92** [comment]
https://github.com/huggingface/lerobot/issues/3076

**Comment by s1lent4gnt** [comment]
> [@s1lent4gnt](https://github.com/s1lent4gnt) i have an update, i started working little by littke. i think first task for me is to refactor learner.py into learner.py and a separate sac.py. i am writing an issue about it. later i will add dsrl.py to this so the learner.py can access it.

Nice work @sezan92 ! but all of this was done in #3075

**Comment by sezan92** [comment]
@s1lent4gnt oh bummer, i want to confirm the workflow as i am confused. should we try to merge in #3075 PR? if yes, initially i thought to target `main` branch. is the DSRL fully merged as planned in this refactor?

**Comment by s1lent4gnt** [comment]
> [@s1lent4gnt](https://github.com/s1lent4gnt) oh bummer, i want to confirm the workflow as i am confused. should we try to merge in [#3075](https://github.com/huggingface/lerobot/pull/3075) PR? if yes, initially i thought to target `main` branch. is the DSRL fully merged as planned in this refactor?

@sezan92 now that #3075 is merged, you can build on top of `main`.

`DSRL` was not merged in that PR; it only refactored the `SAC algorithm` already present on `main`.

**Comment by Asthestarsfalll** [comment]
More RL algorithm like td3,  awac and silri

**Comment by zhoumiaosen** [comment]
Hi! Is this RL contribution request still active? If so, I’d like to contribute a focused beginner-friendly RL training tutorial. Is that a useful scope, and is there a preferred documentation location or example to build on?
